allennlp/nn/util.py

```diff
@@ -72,7 +76,7 @@ def sort_batch_by_length(tensor: torch.autograd.Variable, sequence_lengths: torc
     index_range = Variable(index_range.long())
     _, reverse_mapping = permutation_index.sort(0, descending=False)
     restoration_indices = index_range.index_select(0, reverse_mapping)
-    return sorted_tensor, sorted_sequence_lengths, restoration_indices
+    return sorted_tensor, sorted_sequence_lengths, restoration_indices, permutation_index
```

Comment:
Just a general comment here about the interface, not necessarily a suggestion to change it here:  this is the reason I typically return dicts in situations like this instead of position aware tuples, as it makes modifying the interface much easier.  In this case, adding an additional output required changing all of the places in the code that also called `sort_batch_by_length`.

Reply:
I would tend to agree with this.  Four return values is getting to the point where you probably shouldn't be returning a tuple; something could be made simpler here.  But I don't much like a dict here, either, and making that change would be obnoxious as part of this PR.